BLUEPRINT/utilities/powerlearn.py

```diff
@@ -1,499 +0,0 @@
-                inputs = inputs.drop(var, axis=1)
-                self.x_test = self.x_test.drop(var, axis=1)
-        self.fit(inputs, target)
-        self.r_2 = self.score(inputs, target)
-        self.x_train = inputs
-
-    def _process_xy(self):
-        y_pred, c = self.predict(self.x_test, xc=self.c_test, logged_input=True)
-        return y_pred + c, np.array(self.y_test) + c
-
-    def print_fit(self):
-        """
-        Prints the equation of the LinearLaw and the fitting R^2 score.
-        """
-        varss = self.x_train.columns
-        powers = self.model.coef_
-        k = " + " if self._s != "" else ""
-        eq = "\n" + self.target.name + " = " + self._s + k  # constant variable
-        eq += "{0:.9f}*".format(np.exp(self.model.intercept_))
-        for i in range(len(varss)):
-            eq += "{0:.2f}*{1}".format(powers[i], varss[i])
-            if i != len(varss) - 1:
-                eq += "+"
-        eq += "\nR\u00b2 = {0:.2f}".format(self.r_2)
-        bluemira_print(eq)
-
-
-class PowerLaw(LinearLaw):
-    """
-    Fits a multi-linear regression power law to a dataset.
-    By default optimises the power law by dropping variables one by one
-    and determining if the R**2 metric (%.2f) decreases at all.
-    Trains the power law on 80% of the data (by default), and tests the fit on
-    the remaining 20%. The data is split randomly, meaning that every time the
-    object is run, the power law (and fit quality) can change.
-
-    Parameters
-    ----------
-    dataframe: DataFrame
-        The DataFrame object for which to build a regression Law
-    targets: List[str]
-        The list of column names in the DataFrame which are result columns.
-        These are excluded from the fitting.
-    target: str
-        The result column name in the DataFrame to fit to
-    constant: Union[float, None]
-        The constant (if any) to add to the Law.
-    """
-
-    def __init__(self, dataframe, targets, target, **kwargs):
-        dataframe = np.log(dataframe)
-        super().__init__(dataframe, targets, target, **kwargs)
-
-    def _process_xy(self):
-        # Need to exp it again because of logged_input
-        y_pred, c = self.predict(self.x_test, logged_input=True)
-        return np.exp(y_pred) + c, np.exp(self.y_test)
-
-    def print_fit(self):
-        """
-        Prints the equation of the PowerLaw and the fitting R^2 score.
-        """
-        varss = self.x_train.columns
-        powers = self.model.coef_
-        k = " + " if self._s != "" else ""
-        eq = "\n" + self.target.name + " = " + self._s + k  # constant variable
-        eq += "{0:.9f}*".format(np.exp(self.model.intercept_))
-        for i in range(len(varss)):
-            eq += "{0}^{1:.2f}".format(varss[i], powers[i])
-            if i != len(varss) - 1:
-                eq += "*"
-        eq += "\nR\u00b2 = {0:.2f}".format(self.r_2)
-        bluemira_print(eq)
-
-
-def surface_fit(x, y, z, order: int = 2, n_grid: int = 30):
```

Comment:
I actually used this only the other day to fit some CS fatigue stress / n_cycle data... I will probably need it again in the very near future. Could you move this to somewhere in utilities? I'd say in `tools.py` but this feels specific enough to warrant a new file, like maybe `fit_tools.py` or something. When you rebuild the ML stuff and the power and linear law fitting stuff, they would probably live in the same place. It's a bit of a grey zone between fitting and optimisation, but it is (for me at least) conceptually different and definitely a relatively common activity.

Reply:
I'll do you one better. In #571 I've generalised the function and moved it to a new fit_tools.py for now